Links to source files in go test output are not correct for when testing all pkgs in workspace

Sample project: [gotestbug.zip](https://github.com/Microsoft/vscode-go/files/1908084/gotestbug.zip)

Steps to Reproduce:

1. Load the sample project in VS Code
2. Invoke the Go Test All Packages in Workspace action.
3. go test output looks like the following:

```
Running tool: C:\Go\bin\go.exe test -short -timeout 30s ./...

--- FAIL: TestDoStuff (0.00s)
	c:\Users\gordon.tyler\go\src\gotestbug\stuff_test.go:10: oh noes!
FAIL
FAIL	gotestbug/stuff	0.060s
Error: Tests failed.
```

The link to the source file is incorrect. It should be `c:\Users\gordon.tyler\go\src\gotestbug\stuff\stuff_test.go`.

However, if I open up stuff_test.go and invoke the Go: Test Package action, the go test output contains the correct source file link:

```
Running tool: C:\Go\bin\go.exe test -short -coverprofile=C:\Users\GORDON~1.TYL\AppData\Local\Temp\go-code-cover -timeout 30s gotestbug\stuff

--- FAIL: TestDoStuff (0.00s)
	c:\Users\gordon.tyler\go\src\gotestbug\stuff\stuff_test.go:10: oh noes!
FAIL
coverage: 100.0% of statements
FAIL	gotestbug/stuff	0.067s
Error: Tests failed.
```

--- a/src/testUtils.ts
+++ b/src/testUtils.ts
@@ -174,9 +174,9 @@ export function goTest(testconfig: TestConfig): Thenable<boolean> {
 			const processTestResultLine = (line: string) => {
 				testResultLines.push(line);
 				const result = line.match(packageResultLineRE);
-				if (result) {
+				if (result && currentGoWorkspace) {
 					const packageNameArr = result[2].split('/');
-					const baseDir = path.join(workspaceRoot, ...packageNameArr.slice(1));
+					const baseDir = path.join(currentGoWorkspace, ...packageNameArr);
 					testResultLines.forEach(line => outputChannel.appendLine(expandFilePathInOutput(line, baseDir)));
 					testResultLines.splice(0);
 				}
@@ -184,7 +184,14 @@ export function goTest(testconfig: TestConfig): Thenable<boolean> {
 
 			// go test emits test results on stdout, which contain file names relative to the package under test
 			outBuf.onLine(line => processTestResultLine(line));
-			outBuf.onDone(last => last && processTestResultLine(last));
+			outBuf.onDone(last => {
+				if (last) processTestResultLine(last);
+
+				// If there are any remaining test result lines, emit them to the output channel.
+				if (testResultLines.length > 0) {
+					testResultLines.forEach(line => outputChannel.appendLine(line));
+				}
+			});
 
 			// go test emits build errors on stderr, which contain paths relative to the cwd
 			errBuf.onLine(line => outputChannel.appendLine(expandFilePathInOutput(line, testconfig.dir)));

--- a/src/testUtils.ts
+++ b/src/testUtils.ts
@@ -8,7 +8,7 @@ import path = require('path');
 import vscode = require('vscode');
 import util = require('util');
 import { parseEnvFile, getGoRuntimePath, getCurrentGoWorkspaceFromGOPATH } from './goPath';
-import { getToolsEnvVars, getGoVersion, LineBuffer, SemVersion, resolvePath, getCurrentGoPath } from './util';
+import { getToolsEnvVars, getGoVersion, LineBuffer, SemVersion, resolvePath, getCurrentGoPath, getWorkspaceFolderPath } from './util';
 import { GoDocumentSymbolProvider } from './goOutline';
 import { getNonVendorPackages } from './goPackages';
 
@@ -150,6 +150,8 @@ export function goTest(testconfig: TestConfig): Thenable<boolean> {
 			args.push(testconfig.dir.substr(currentGoWorkspace.length + 1));
 		}
 
+		const workspaceRoot = getWorkspaceFolderPath(vscode.Uri.file(testconfig.dir));
+
 		targetArgs(testconfig).then(targets => {
 			let outTargets = args.slice(0);
 			if (targets.length > 2) {
@@ -166,9 +168,25 @@ export function goTest(testconfig: TestConfig): Thenable<boolean> {
 			const outBuf = new LineBuffer();
 			const errBuf = new LineBuffer();
 
-			outBuf.onLine(line => outputChannel.appendLine(expandFilePathInOutput(line, testconfig.dir)));
-			outBuf.onDone(last => last && outputChannel.appendLine(expandFilePathInOutput(last, testconfig.dir)));
+			const packageResultLineRE = /^(ok|FAIL)[ \t]+(.+?)[ \t]+([0-9\.]+s|\(cached\))/; // 1=ok/FAIL, 2=package, 3=time/(cached)
+			const testResultLines: string[] = [];
 
+			const processTestResultLine = (line: string) => {
+				testResultLines.push(line);
+				const result = line.match(packageResultLineRE);
+				if (result) {
+					const packageNameArr = result[2].split('/');
+					const baseDir = path.join(workspaceRoot, ...packageNameArr.slice(1));
+					testResultLines.forEach(line => outputChannel.appendLine(expandFilePathInOutput(line, baseDir)));
+					testResultLines.splice(0);
+				}
+			};
+
+			// go test emits test results on stdout, which contain file names relative to the package under test
+			outBuf.onLine(line => processTestResultLine(line));
+			outBuf.onDone(last => last && processTestResultLine(last));
+
+			// go test emits build errors on stderr, which contain paths relative to the cwd
 			errBuf.onLine(line => outputChannel.appendLine(expandFilePathInOutput(line, testconfig.dir)));
 			errBuf.onDone(last => last && outputChannel.appendLine(expandFilePathInOutput(last, testconfig.dir)));
 
@@ -230,4 +248,3 @@ function targetArgs(testconfig: TestConfig): Thenable<Array<string>> {
 	}
 	return Promise.resolve([]);
 }
-

